### ISSUE TITLE
Messageboxes and dialogs

### ISSUE THREAD

**Issue by rdbende** [tukaan/tukaan#38]
TODO:
- [ ] Messageboxes ([Sun-Valley-messageboxes](https://github.com/rdbende/Sun-Valley-messageboxes))
- [x] Filedialogs
  - [x] Open file
  - [x] Save file
  - [x] Select directory
- [ ] Color selection
- [ ] Input dialog
  - [ ] String
  - [ ] Number
  - [ ] Slider
- [ ] Calendar dialog
- [ ] About dialog
- [ ] Task dialog
- [ ] Progress dialog
- [ ] Notifications
- [ ] Credentials dialog
- [ ] Print dialog
- [ ] Custom dialog API

Implementation:
- [x] Use `KDialog` if available
- [x] Use `Zenity` if available
- [x] Use `notify-send` if available

**Comment by sumeshir26** [comment]
Where can I get these tkinter message boxes?

**Comment by rdbende** [comment]
[Right here](https://github.com/rdbende/Sun-Valley-messageboxes)

**Comment by Moosems** [comment]
@rdbende How can I make a custom messagebox like those?

**Comment by Moosems** [comment]
Can we try and add the Mac printer dialog?

**Comment by rdbende** [comment]
> It would also be nice to have a built-in way to make dialogs

Yup, it isn't in the TODO, but actually I've implemented it.

> Can we try and add the Mac printer dialog?

Definitely, amazing good idea! I'll have to do some research, because it's probably not that straightforward, but it would be nice to have a print dialog (not only on Mac :)

**Comment by Moosems** [comment]
> Definitely, amazing good idea! I'll have to do some research, because it's probably not that straightforward, but it would be nice to have a print dialog (not only on Mac :)

Yeah its not built into tk and haven't been able to find anything yet but that would be an amazing add

**Comment by rdbende** [comment]
> https://stackoverflow.com/questions/72370788/bring-up-macos-print-file-dialog

@Moosems Here are your requested links:
- https://github.com/tcltk/tk/blob/main/library/print.tcl
- https://github.com/tcltk/tk/blob/main/macosx/tkMacOSXPrint.c

I haven't read through them, just found the files, but it looks like there is indeed going to be some printing stuff in Tk8.7

[https://wiki.tcl-lang.org/page/Tcl/Tk+Printing+Support](https://wiki.tcl-lang.org/page/Tcl%2FTk+Printing+Support)

**Comment by Moosems** [comment]
That's awesome! You know that TRCK on stackoverflow is me right?

**Comment by rdbende** [comment]
Yes :)

**Comment by Moosems** [comment]
Haha I just thought it was kinda funny how you had pulled up my question

**Comment by Moosems** [comment]
How can I call messageboxes?

**Comment by rdbende** [comment]
In theory you could do this:
```python
from tukaan.dialogs import FileDialog


dir = FileDialog.get_directory()

file_to_open = FileDialog.get_open_filenames()
file_to_save = FileDialog.get_save_filenames()


with FileDialog.get_open_file() as file:
    print(file.read())

with FileDialog.get_save_file() as file:
    file.write("asdf")
```
but something is broken, so you can't ATM.

**Comment by Moosems** [comment]
Oh ok. How can I help get them working?

**Comment by rdbende** [comment]
At the moment I don't you could help, because I don't know what causes the error either. But if you can figure it out, then...

**Comment by Moosems** [comment]
Ok. Where is the code for it located?

**Comment by Moosems** [comment]
Thanks

**Comment by rdbende** [comment]
Windows toast notifications: https://github.com/mohabouje/WinToast
